dust-tt/dust
# [Vaults] Restrict access to conversations and messages

## Description

Fixes https://github.com/dust-tt/dust/issues/6903

Left a few comments on the PR for reviewers to check.

Now that agents have ACLs, conversations and messages should be restricted to users that have access to the related assistants.

Restriction is implemented on conversations by: 
- checking if any of its agent messages shouldn't be accessed by the user. ACLs are naturally present on agent messages via their attached agent configuration's groupIds.
- in cases where we don't have access to all the messages of a conversation, checking conversation access via a new function `canAccessConversation`,  relying on conversation participants---that will now return an error if one of the agents or more is not accessible to the user;

Summary: 
- checked all references to AgentMessage and Message models (details below) and acted where needed
- changed getConversation to a Result with ConversationPermissionError / ConversationNotFoundError
  - wrapped conversation api errors in a dedicated function, drying quite a bit
- did the same with getConversationWithoutContent, further drying
- double-check: all endpoints, and references to Conversation


## Risk
### Latency
Some calls become more heavy due to the addition of a `canAccessConversation` call relying on `fetchConversationParticipants` and as such making 2 additional database queries. The calls are indexed and query results are ~small volume, but still worth monitoring. This is the cost of checking per-conversation in addition to per-messages.

### Testing
Test scenarios:
- with allowed agents, all below should work
  - UX 
	- start a convo :heavy_check_mark: 
	- read a convo :heavy_check_mark: 
	- post a message to a convo :heavy_check_mark: 
	- destroy a convo :heavy_check_mark: 
  - API v1
	- start a convo, 
	- read a convo, 
	- post a message to a convo, 
	- destroy a convo
- with unallowed agents, all below should fail
  - start a convo (only API v1)
  - read a convo (both UX and v1) :heavy_check_mark: 
  - post a message to a convo (only v1) 
  - destroy a convo (only v1)

## Deploy Plan
test on front-edge
front
